book/python/writing-your-own-stateful-application.md

```diff
@@ -1,96 +1,70 @@
 
-As with the Reverse Word example, we will list the components required:
+As with the stateless Alerts example, we will list the components required:
 
-* Input decoding
 * Output encoding
-* Computation for adding votes
-* State objects
-* State change management
+* Computation for checking transactions
+* State object
 
-### StateComputation
+### State Computation
 
-The state computation here is fairly straightforward: given a data object and a state object, update the state with the new data, and return some data that tells Wallaroo what to do next.
+The state computation here is fairly straightforward: given a data object and a state object, update the state with the new data, and, based on a simple condition, determine whether we return some data representing an alert message.
 
+Our input type represents a transaction:
 ```python
-@wallaroo.state_computation(name='add votes')
-def add_votes(data, state):
-    state.update(data)
-    return (state.get_votes(data.letter), True)
+class Transaction(object):
+    def __init__(self, user, amount):
+        self.user = user
+        self.amount = amount
 ```
 
-Let's dig into that tuple that we are returning:
+We use these inputs to update state, and check whether the running total crosses certain thresholds, at which point we generate an alert:
 
 ```python
-(state.get_votes(data.letter), True)
+@wallaroo.state_computation(name="check transaction total", state=TransactionTotal)
+def check_transaction_total(transaction, state):
+    state.total = state.total + transaction.amount
+    if state.total > 2000:
+        return DepositAlert(transaction.user, state.total)
+    elif state.total < -2000:
+        return WithdrawalAlert(transaction.user, state.total)
 ```
 
-The first element, `state.get_votes(data.letter)`, is a message that we will send on to our next step. In this case, we will be sending information about votes for this letter on to a sink. The second element, `True`, is to let Wallaroo know if we should store an update for our state. By returning `True`, we are instructing Wallaroo to save our updated state so that in the event of a crash, we can recover to this point. Being able to recover from a crash is a good thing, so why wouldn't we always return `True`? There are two answers:
+### State
 
-1. Your state computation might not have updated the state, in which case saving its state for recovery is wasteful.
-2. You might only want to save after some changes. Saving your state can be expensive for large objects. There's a trade-off that can be made between performance and safety.
-
-### State and StateBuilder
-
-The state for this application is two-tiered. There is a vote count for each letter:
+The state for this application keeps track of a running total of transactions:
 
 ```python
-class Votes(object):
-    def __init__(self, letter, votes):
-        self.letter = letter
-        self.votes = votes
+class TransactionTotal(object):
+    total = 0
```

Comment:
This should be:

```python
class TransactionTotal(object):
    def __init__(self):
        self.total = 0
```

Reply:
Do you say this because it's not idiomatic to do it the compact way?  It does work.

Reply:
It creates a class variable that's shared by all instances.